### DISCUSSION TITLE
add global nets routing to module checkpoints

### DISCUSSION THREAD

**Discussion by GBrics** [Xilinx/RapidWright#434] [Q&A] [answered]
Hello, 

I'm currently facing an issue when exporting individual checkpoint files for each module of my design:
when I read the checkpoint file, I can notice that the clock nets (originating from a MMCM outside the module) are not routed. Only intraSlice routing are depicted. 

Is there anyway to force the clock nets (and more generally any net that cross the module boundaries) to be included in the checkpoint files ? 

Thank you in advance!

GBrick

**Comment by clavin-xlnx** [comment, marked as answer]
Hi GBrick,

When you describe "exporting individual checkpoint files", do you mean running `write_checkpoint -cell` in Vivado?  What you describe seems to be the behavior of that Tcl proc.  I don't think there is a way in Vivado to preserve the cross-boundary nets.  

You can try to restore the clock routing (either fully or in part) by copying it from the original DCP where the clock is routed and pasting it into the module-based DCP.  You can do this either in RapidWright or Vivado.  In RapidWright, you'd need to load both DCPs and then get the `Net` object with the routing you want to copy as well as the destination `Net` and copy the `PIP` objects over.  For example:

```java
Design srcDesign = Design.readCheckpoint(...);
Design dstDesign = Design.readCheckpoint(...);
...
dstDesign.getNet(destNetName).setPIPs(srcDesign.getNet(srcNetName).getPIPs());
```



In Vivado, a similar operation could be performed using:
```tcl
set pips [get_pips -of $your_src_net]
set_property FIXED_ROUTE $pips $your_dest_net
```

**Comment by GBrics** [comment]
Thank you for your answer.

Indeed, the issue was occurring when running the ``write_checkpoint -cell`` command from Vivado.

I couldn't solve the problem with the second solution (directly from Vivado) as multiple errors were reported indicating the absence of some PIPs when copying them to the destination net (I couldn't figure out why these errors were reported).

However, by using the first solution with RapidWright, all the PIPs from the source net have been properly copied to the destination node. By writing back a checkpoint file with the corrected design, I was able to confirm the effectiveness of the solution from Vivado GUI.

Thank you very much for your precious help!

GBrick
